scripts/generate_refcats_gen3.py

```diff
@@ -0,0 +1,248 @@
+#!/usr/bin/env python
+# This file is part of ap_verify_hits2015.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+# Script for ingesting standard refcats that cover this dataset's fields.
+# Running this script allows for updates to the refcats to be incorporated
+# into the dataset.
+
+import argparse
+import logging
+import sys
+
+from esutil import htm
+from astropy.coordinates import SkyCoord
+
+import lsst.log
+from lsst.daf.butler import Butler, CollectionType, DatasetRef, DatasetType, FileDataset
+
+
+logging.basicConfig(level=logging.INFO, stream=sys.stdout)
+lsst.log.configure_pylog_MDC("DEBUG", MDC_class=None)
+
+
+########################################
+# Fields and catalogs to process
+
+FIELDS = [SkyCoord("10h00m28.800s", "+02d12m36.00s"),  # field 26 (2014:04)
+          SkyCoord("10h20m28.800s", "-06d31m12.00s"),  # field 40 (2014:10)
+          SkyCoord("10h21m52.800s", "-04d57m00.00s"),  # field 42 (2014:09)
+          ]
+FIELD_RADIUS = 2.0  # degrees
+
+# Key is catalog name in source repo, value is name in ap_verify_hits2015.
+REFCATS = {"gaia_dr2_20200414": "gaia",
+           "ps1_pv3_3pi_20170110": "panstarrs",
+           }
+
+
+########################################
+# Command-line options
+
+def _make_parser():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("-b", dest="src_dir", default="/repo/main",
+                        help="Refcat source Butler repo, defaults to '/repo/main'.")
+    parser.add_argument("-i", dest="src_collection", default="refcats",
+                        help="Refcat source collection, defaults to 'refcats'.")
+    return parser
+
+
+args = _make_parser().parse_args()
+
+
+########################################
+# Identify all required shards
+
+HTM_LEVEL = 7
+
+
+def _get_shards(centers, radius):
+    """Return all shards overlapping a set of fields.
```

Comment:
👍 for using `sphgeom.HtmPixelization.envelope()`, 👎 for trying to create some kind of universal solution for everybody. The only reason this is even a function is to keep the namespace uncluttered.

Reply:
A question about using `HtmPixelization`: looking at the docs it sounds like this does not guarantee that it would return pixels of `HTM_LEVEL`, but reserves the right to return equivalent higher-level pixels. Is this correct? Since I'm using them as refcat IDs rather than doing a true spatial query, any "merged" pixels are unusable.

Reply:
I don't know the answer on that count. @TallJimbo would be interested and might know, since he was working on DM-14434.

Reply:
I believe `HtmPixelization` is guaranteed to return pixels at the requested level, but it may perform the computation at a coarser level and then just bit-shift to the requested level, yielding a conservative approximation to ideal solution at the requested level.

Reply:
...and yes, we do want to stop using esutil.

Reply:
FWIW, the work I started on that was actually on [DM-30275](https://jira.lsstcorp.org/browse/DM-30275), and the last comment there explains how far I got; it's not really actively in progress anymore, just because my assumption that it was doable quickly turned out to be false.  I also just closed DM-14434 as a duplicate.

Reply:
I've managed to do the conversion, and eyeballed that I got the same shards out the other end. The fact that `HtmPixelization` returns ranges rather than individual pixels means that a lot of the range-conversion logic from the original script is no longer needed.

Reply:
"universal solution for everybody" - I mean, I've needed exactly this function myself a number of times, when extracting refcat shards for other test data, which is why sphgeom seems like a good place for it to me. "I have a list of circles, give me all the shards that overlap them" seems generally useful.

Reply:
Does this also apply to the new implementation? There are a few design decisions here (using `astropy.SkyCoord` instead of `sphgeom.LonLat` as input, or returning ranges instead of explicit shards) that were very much to solve this script's problems and might not make sense in general.

Reply:
Lets ticket it, and we can decide on an interface there.